drivers: sensor: bosch: add support for bmp180

jovicjakov opened the pull request:
This PR aims to add driver support for the Bosch BMP180 digital pressure and temperature sensor.

jovicjakov committed:
drivers: sensor: bosch: add support for bmp180

The Bosch BMP180 is a sensor that measures pressure and temperature.

Signed-off-by: Jakov Jović <jovic.jakov@outlook.com>

jovicjakov committed:
dts: bindings: sensor: add bindings for bmp180

Add bindings for the Bosch BMP180 pressure and temperature sensor.

Signed-off-by: Jakov Jović <jovic.jakov@outlook.com>

jovicjakov committed:
tests: build_all: add entries for bmp180

Add the new BMP180 drivers to the test suite.

Signed-off-by: Jakov Jović <jovic.jakov@outlook.com>

kartben commented on samples/sensor/bmp180/src/main.c:
please just drop this sample altogether as it's not adding anything useful compared to e.g. sensor_shell sample which folks can use to quickly test their sensor.

jovicjakov replied:
I agree, I'll remove the commit

kartben reviewed (dismissed):
some early comments. thanks for this new driver!

jovicjakov commented on drivers/sensor/bosch/bmp180/bmp180.c:
```diff
@@ -0,0 +1,297 @@
+	uint8_t buf[2];
+
+	if (i2c_reg_write_byte_dt(&config->i2c, BMP180_REG_CTRL_MEAS, BMP180_CMD_READ_TEMP) < 0) {
+		LOG_ERR("Failed to start the temperature measurement.");
```
will do

jovicjakov commented on drivers/sensor/bosch/bmp180/bmp180.c:
```diff
@@ -0,0 +1,297 @@
+		break;
+	case SENSOR_CHAN_ALTITUDE:
+		val->val1 = data->alt;
+		val->val2 = 0;
```
good catch, thanks

ubieda reviewed (commented):
Thanks for your contribution! Small suggestion, otherwise looks good!

jovicjakov commented on drivers/sensor/bosch/bmp180/bmp180.c:
```diff
@@ -0,0 +1,294 @@
+		return -EIO;
+	}
+
+	data->uncomp_press = ((int32_t)buf[0] << 16) | ((int32_t)buf[1] << 8) | buf[2];
```
ofc, I'll refactor

jovicjakov commented:
@decsny, when you have time, could you take another look at this? Thanks!

MaureenHelm commented on drivers/sensor/bosch/bmp180/bmp180.c:
```diff
@@ -0,0 +1,294 @@
+	.channel_get = bmp180_channel_get,
+};
+
+int bmp180_init(const struct device *dev)
```
```suggestion
static int bmp180_init(const struct device *dev)
```

jovicjakov replied:
Thanks, resolved

MaureenHelm commented on drivers/sensor/bosch/bmp180/bmp180.c:
```diff
@@ -0,0 +1,294 @@
+	const struct bmp180_config *config = dev->config;
+	uint8_t id = 0U;
+
+	if (!device_is_ready(config->i2c.bus)) {
```
```suggestion
	if (!i2c_is_ready_dt(&config->i2c)) {
```

jovicjakov replied:
Thanks, resolved

MaureenHelm commented on drivers/sensor/bosch/bmp180/Kconfig:
```diff
@@ -0,0 +1,12 @@
+	select I2C
+	default y
+	help
+		Enable the driver for the Bosch BMP180 pressure sensor.
```
```suggestion
	  Enable the driver for the Bosch BMP180 pressure sensor.
```

jovicjakov replied:
Thanks, resolved

MaureenHelm reviewed (changes requested):
Some nits, but otherwise looks good. Thanks for your contribution!

jovicjakov commented:
@MaureenHelm Thank you for the review! I've fixed the issues you pointed out. Please take another look when you have a chance.  :)

ubieda commented:
@MaureenHelm it does appear everything's addressed. There's a conflict resolution required but seems trivial to fix.

kartben commented:
> @MaureenHelm it does appear everything's addressed. There's a conflict resolution required but seems trivial to fix.

@ubieda  feel free to directly push to the branch to save some time :)

ubieda reviewed (dismissed):
Re-approving after rebasing

aescolar commented:
Bulk changing the milestone from everything which did not make it to the v3.7.0 freeze deadline
and which is not labeled as a bug or documentation (or does not appear clearly as only containing
a bug fix or a documentation change).
If you disagree about this change please add the milestone again.

Please note that until rc2 only PRs containing bug fixes, docs and tests can be merged.
After that and until release only bug fixes and docs changes.

Exceptions require approval by the release team and a vote by the TSC.
https://docs.zephyrproject.org/latest/project/release_process.html#stabilization-phase

aescolar commented:
@jovicjakov rebase needed

kartben commented:
rebased to resolve conflict

aescolar commented:
@MaureenHelm please re-review

mariopaja commented:
This driver can also be used on BMP085. Maybe it can be documented somewhere since https://github.com/zephyrproject-rtos/zephyr/pull/67612 was blocked

ruehlchris commented on drivers/sensor/bosch/bmp180/bmp180.c:
```diff
@@ -0,0 +1,294 @@
+		val->val2 = data->temp % 10 * 100000;
+		break;
+	case SENSOR_CHAN_PRESS:
+		/* Pressure is calculated in steps of 1 Pa */
```
CHAN PRESSURE for the Senor API should be in kPa not Pa.

ruehlchris commented on drivers/sensor/bosch/bmp180/bmp180.c:
```diff
@@ -0,0 +1,294 @@
+	int32_t b6 = b5 - 4000;
+
+	x1 = (data->b2 * ((b6 * b6) >> 12)) >> 11;
+	x2 = (data->ac2 * b6) >> 11;
```
you might want to simplify from the original data-sheet example to 
x1 = ((b2 * (b6 * b6)>>12))>>11;
x1 = ((b2 * b6 * b6) / pow(2,12) )/ pow(2,11)
x1 = (b2 * b6 *b6) (pow(2,12) * pow(2,11))
x1 = (b2 * b6 * b6) / 0x800000;  << much

ruehlchris commented on drivers/sensor/bosch/bmp180/bmp180.c:
```diff
@@ -0,0 +1,294 @@
+	} else {
+		p = (b7 / b4) * 2;
+	}
+
```
I wondered the if (b7< xx) statement here from the datasheet example
math wise it is exactly same  only a data overflow can happen if b7 is too big
so b7/b4 *2 would always be safe and no need for a if()

ruehlchris commented:
Hi @jovicjakov ,  it is a pity that I missed you PR request and did my own implementation. 
Not a problem but wast a bit of time.  I saw in your implementation a lot of the original code from the datasheet which are not optimal (IMHO). I add some comments to your code lately. 
As well, you exit with an error if the command ready bit(5) not 0 rather wait for the convention is ready.
The Sensor API wants the Pressure return as kPa your implementation return Pa.
Please have a look into my PR and use it to your benefit.  #77182

jovicjakov commented:
Hi everyone,

I just wanted to inform you that I have started an internship recently, and due to the terms of my contract, I am not allowed to contribute to open-source projects, not at this moment at least. I won't be able to continue working on this PR or addressing the remaining comments.

If any further changes are required, I may be unable to make them myself. I would be grateful if someone else could take over from here. Thanks!
